akka-docs/src/main/paradox/typed/reliable-delivery.md

```diff
@@ -0,0 +1,358 @@
+subscribes to the same key to find active workers. In this way workers can be dynamically
+added or removed from any node in the cluster.
+
+The work manager (producer) actor will start the flow by sending a `WorkPullingProducerController.Start`
+message to the `WorkPullingProducerController`.
+
+The `WorkPullingProducerController` sends `RequestNext` to the producer, which is then allowed
+to send one message to the `WorkPullingProducerController`.
+Thereafter the producer will receive a new `RequestNext` when it's allowed to send one more message.
+It will send a new `RequestNext` when there are demand from any worker.
+It's possible that all workers with demand are deregistered after the `RequestNext` is sent and before
+the actual messages is sent to the `WorkPullingProducerController`. In that case the message is
+buffered and will be delivered when a new worker is registered or when there is new demand.
+
+The producer and `WorkPullingProducerController` actors are supposed to be local so that these messages are
+fast and not lost.
+
+Similarly, on the consumer side the destination consumer actor will start the flow by sending an
+initial `ConsumerController.Start` message to the `ConsumerController`.
+
+Received messages from the producer are wrapped in `ConsumerController.Delivery` when sent to the consumer,
+which is supposed to reply with `ConsumerController.Confirmed` when it has processed the message.
+Next message is not delivered until the previous is confirmed. More messages from the producer that arrive
+while waiting for the confirmation are stashed by the `ConsumerController` and delivered when previous
+message was confirmed.
+
+The consumer and the `ConsumerController` actors are supposed to be local so that these messages are fast
+and not lost.
+
+Many unconfirmed messages can be in flight between the `WorkPullingProducerController` and each
+`ConsumerController`, but it is limited by a flow control window. The flow control is driven by the
+consumer side, which means that the `WorkPullingProducerController` will not send faster than the
+demand requested by the workers.
+
+Example of image converter worker (consumer):
+
+Scala
+:  @@snip [WorkPullingDocExample.scala](/akka-cluster-sharding-typed/src/test/scala/docs/delivery/WorkPullingDocExample.scala) { #imports #consumer }
+
+and image converter job manager (producer):
+
+Scala
+:  @@snip [WorkPullingDocExample.scala](/akka-cluster-sharding-typed/src/test/scala/docs/delivery/WorkPullingDocExample.scala) { #imports #producer }
+
+FIXME Java example
+
+Note how the `ActorRef` in the `Start` messages are constructed as message adapters to map the
+`RequestNext` and `Delivery` to the protocol of the producer and consumer actors respectively.
+
+### Work pulling delivery semantics
+
+For work pulling the order of the messages should not matter, because each message is routed randomly
+to one of the workers with demand.
```

Comment:
Adding "and can therefore be processed in any order" or something along those lines would make it more clear?